test/cpp/asyncprogressqueueworkerstream.cpp

```diff
@@ -0,0 +1,80 @@
+/*********************************************************************
+ * NAN - Native Abstractions for Node.js
+ *
+ * Copyright (c) 2017 NAN contributors
+ *
+ * MIT License <https://github.com/nodejs/nan/blob/master/LICENSE.md>
+ ********************************************************************/
+
+#include <nan.h>
+
+using namespace Nan;  // NOLINT(build/namespaces)
+
+// Custom data type: This serves as an example of how external
+// libraries could be hooked in, populate their objects and send them to JS.
+struct data_t {
+  int index;
+  int data;
+};
+
+template<typename T>
+class ProgressQueueWorker : public AsyncProgressQueueWorker<T> {
+ public:
+  ProgressQueueWorker(
+      Callback *callback
+    , Callback *progress
+    , int iters)
+    : AsyncProgressQueueWorker<T>(callback), progress(progress)
+    , iters(iters) {}
+
+  ~ProgressQueueWorker() {
+    delete progress;
+  }
+
+  void Execute (
+    const typename AsyncProgressQueueWorker<T>::ExecutionProgress& progress) {
+    data_t data;
+    for (int i = 0; i < iters; ++i) {
+      data.index = i;
+      data.data = i * 2;
+      progress.Send(&data, 1);
+    }
+  }
+
+  void HandleProgressCallback(const T *data, size_t count) {
+    HandleScope scope;
+    v8::Local<v8::Object> obj = Nan::New<v8::Object>();
+    Nan::Set(
+      obj,
+      Nan::New("index").ToLocalChecked(),
+      New<v8::Integer>(data->index));
+    Nan::Set(
+      obj,
+      Nan::New("data").ToLocalChecked(),
+      New<v8::Integer>(data->data));
+
+    v8::Local<v8::Value> argv[] = { obj };
+    progress->Call(1, argv);
+  }
+
+ private:
+  Callback *progress;
+  int iters;
+};
+
+NAN_METHOD(DoProgress) {
+  Callback *progress = new Callback(info[1].As<v8::Function>());
+  Callback *callback = new Callback(info[2].As<v8::Function>());
```

Comment:
Future improvement: update all instances of `val.As<v8::Function>()` to `To<v8::Function>(val).ToLocalChecked()` in the test suite.

Reply:
should we turn this (and the `vector<T>.swap()` idea from pr 692: https://github.com/nodejs/nan/pull/692#discussion_r142992578 ) into github issues so that we dont forget, and maybe one of us (or maybe a noob looking to get involved) will have a place to find ideas for improvements?